cpp/daal/src/data_management/roc_auc_score.cpp

```diff
@@ -0,0 +1,149 @@
+        for (size_t i = 0; i < blockSize; ++i)
+        {
+            predict[blockBegin + i].value = testPredictionPtr[blockBegin + i];
+            predict[blockBegin + i].index = blockBegin + i;
+        }
+    });
+
+    parallel_sort<FPType>(predict.get(), predict.get() + nElements);
+
+    size_t rank            = 1;
+    size_t elementsInBlock = 1;
+    size_t i               = 0;
+    while (i < nElements)
+    {
+        size_t j = i;
+        while ((j < (nElements - 1)) && (predict[j].value == predict[j + 1].value))
+        {
+            j++;
+        }
+        elementsInBlock = j - i + 1;
+
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        for (size_t j = 0; j < elementsInBlock; ++j)
+        {
+            const size_t idx   = predict[i + j].index;
+            predictedRank[idx] = static_cast<FPType>(rank) + ((static_cast<FPType>(elementsInBlock) - FPType(1.0)) * FPType(0.5));
+        }
+        rank += elementsInBlock;
+        i += elementsInBlock;
+    }
+    return s;
+}
+
+template <typename FPType>
+DAAL_EXPORT void calculateRankData(FPType * predictedRank, const NumericTablePtr & testPrediction, const size_t nElements)
+{
+#define DAAL_CALC_RANK_DATA(cpuId, ...) calculateRankDataImpl<FPType, cpuId>(__VA_ARGS__);
+
+    DAAL_DISPATCH_FUNCTION_BY_CPU_SAFE(DAAL_CALC_RANK_DATA, predictedRank, testPrediction, nElements);
+
+#undef DAAL_CALC_RANK_DATA
+}
+
+template DAAL_EXPORT void calculateRankData<float>(float * predictedRank, const NumericTablePtr & testPrediction, const size_t nElements);
+template DAAL_EXPORT void calculateRankData<double>(double * predictedRank, const NumericTablePtr & testPrediction, const size_t nElements);
+
+template <typename FPType, daal::CpuType cpu>
+services::Status rocAucScoreImpl(const FPType * const predictedRank, const NumericTablePtr & truePrediction, const size_t nElements, FPType & score)
+{
+    services::Status s;
+    ReadRows<FPType, cpu> truePredictionBlock(truePrediction.get(), 0, 1);
+    const FPType * const truePredictionPtr = truePredictionBlock.get();
+    DAAL_CHECK_BLOCK_STATUS(truePredictionBlock);
+    FPType sum = FPType(0);
+
+    PRAGMA_IVDEP
+    PRAGMA_VECTOR_ALWAYS
+    for (size_t i = 0; i < nElements; ++i)
+    {
+        sum += truePredictionPtr[i];
+    }
+    const FPType nPos = sum;
+    const FPType nNeg = static_cast<FPType>(nElements) - nPos;
+
+    FPType filteredRankSum = FPType(0);
+
+    PRAGMA_IVDEP
+    PRAGMA_VECTOR_ALWAYS
+    for (size_t i = 0; i < nElements; ++i) // parallel this
+    {
+        if (truePredictionPtr[i] == FPType(1))
+        {
+            filteredRankSum += predictedRank[i];
+        }
+    }
+
+    score = (filteredRankSum - (nPos * (nPos + FPType(1.0)) * FPType(0.5))) / (nPos * nNeg);
+    return s;
+}
+
+template <typename FPType>
+DAAL_EXPORT FPType rocAucScore(const FPType * const predictedRank, const NumericTablePtr & truePrediction, const size_t nElements)
```

Comment:
I think parameter `nElements` unnecessary, since you can get this information from the tables